### PR TITLE
Update Dynamic CP to support sequence packing, MXFP8 and Mamba

### PR DESCRIPTION
**Pull request by parthmannan**
# What does this PR do ?
Note: Hybrid CP is referred to as Dynamic CP. Naming update in code will be a separate PR.

This PR updates Dynamic CP to support performance and functional improvements - 
1. Allows for sequence packing along with Dynamic CP. This is incredibly useful when very short sequences are part of the dataset and are individually not sufficient to saturate GPU compute even at CP=1. (credit: @xiaoyao0115)
2. Adds support for Mamba and MTP
3. Adds required padding support inside data scheduler to support HybridEP and MXFP8.
4. Fixes seq_load_balancing_loss and mtp_loss calculations when using Dynamic CP. Regular training uses `tp_cp_group` which is incorrect for Dynamic CP as we may have multiple sub-sequences running independently inside a TPxCP group. This PR fixes the load balancing loss calculation used for backward and approximately fixes the calculation for logging. This is because the logging logic would require additional reductions and hamper performance. We will study that further and improve in future.
5. This also removes the SFT loss calculation discrepancy between training and eval. In eval, the old `if args.sft` code path was still alive which has been removed for training loss and this brings both paths to equivalent state. In the old path, for SFT, we were calculating loss per sequence and then doing reduction (i.e. `(loss_0 / total_valid_tokens_0) + (loss_1 / total_valid_tokens_1)`) instead of `total_loss / total_valid_tokens` (i.e. `(loss_0 + loss_1) / (total_valid_tokens_0 + total_valid_tokens_1)`).

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
6. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3791/changes/d9772a3861c2193b32b39bb9a2ca3e774300cf76

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3791/changes/8b80f1ae17088951cdb2d34a82acd7b221b75505

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3791/changes/ba2a4548c2d3d87057bb60ad1d16a1c47cd93f21

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/3791/changes/847483c3fe70d9d0fa104c486a2f844ea7d2ef94

File: megatron/core/datasets/data_schedule.py [hunk 1]
```diff
@@ -264,6 +265,182 @@ def unpack_batch(self, batch):
...
+            tokens = torch.cat(
+                [tokens, torch.zeros(pad_len, dtype=tokens.dtype, device=tokens.device)]
+            )
+            labels = torch.cat(
+                [labels, torch.zeros(pad_len, dtype=labels.dtype, device=labels.device)]
+            )
+            loss_mask = torch.cat(
+                [loss_mask, torch.zeros(pad_len, dtype=loss_mask.dtype, device=loss_mask.device)]
+            )
+            position_ids = torch.cat(
```
[the remark this answers is not in the document]
Comment (parthmannan) [reply megatron/core/datasets/data_schedule.py:344]: Yes you are right. We can just set `pad_granularity` to 1 after the first block and check `if pad_granularity > 1:` to apply the second block of padding.

File: megatron/core/datasets/data_schedule.py [hunk 2]
```diff
@@ -264,6 +265,187 @@ def unpack_batch(self, batch):
...
+            ):
+                # HybridEP requires MAX_NUM_OF_TOKENS_PER_RANK to be divisible
+                # by NUM_OF_TOKENS_PER_CHUNK (128)
+                sharded_pad_granularity = 128
+
+            sharded_tensor_shape = total_seq_len // (local_cp_size * tp_size)
+            mod_token_count = sharded_tensor_shape % sharded_pad_granularity
+            sharded_pad_len = 0
+            if mod_token_count != 0:
+                sharded_pad_len = (sharded_pad_granularity - mod_token_count) * (local_cp_size * tp_size)
```
[the remark this answers is not in the document]
Comment (parthmannan) [reply megatron/core/datasets/data_schedule.py:310]: Fixed it. Thank you!
